docs: improve versioning documentation (#9017)
renovatebot/renovate: docs/usage/modules/versioning.md

The "versioning" is different for each package manager, because different package managers use different versioning schemes.
For example, `npm` uses`1.0.0-beta.1` and `pip` uses `1.0.0b1`.

> ## Why you might need to tinker with the versioning

Reviewer: ```suggestion
## Why you might need to manually configure versioning
```

> Renovate tries to do the right thing out-of-the-box most of the time.

Reviewer: ```suggestion
Renovate interprets versions correctly out-of-the-box most of the time.
```

> It's impossible to support **all** versioning schemes, so sometimes you need to tell the bot what versioning scheme it should use.

Reviewer: ```suggestion
It's impossible to automatically detect **all** versioning schemes, so sometimes you need to tell the bot what versioning scheme it should use.
```

You can manually configure/override the `versioning` value for a particular dependency.
You generally won't need to override the defaults for ecosystems which enforce a strict version scheme like `npm`.

> Configuring or overriding the default `versioning` can be helpful for ecosystems like Docker or helm-values, where versioning is barely a "convention".

Reviewer: ```suggestion
Configuring or overriding the default `versioning` can be particularly helpful for ecosystems like Docker/Kubernetes/Helm, where versioning is barely a "convention".
```

## General concepts behind overriding versioning

> TODO:
> 
> - Explain how/why to use `packageRules`
> - Explain what kind of things you can set with `packageRules`
> - Explain what are the most common proporties you will need in `packageRules` to get what you want

Reviewer: ```suggestion
- Although you can reconfigure versioning per-manager or per-datasource, you are unlikely to ever to need to do this
- More commonly you would need to configure `versioning` for individual packages or potentially package patterns
- The best way to do this is with `packageRules`, with a combination of `matchManagers`, `matchDatasources`, `matchPackageNames` and `matchPackagePatterns`
```

> ## Examples of commonly needed versioning overrides
> 

Reviewer: ```suggestion
## Examples of versioning overrides

### Overriding Docker versioning to use a versioning specific for a package

```json
{
  "packageRules": [
    {
      "matchDatasources": ["docker"],
      "matchPackageNames": ["python"],
      "versionScheme": "pep440"
    }
  ]
}
```

### Using a custom regex versioning scheme

```json
{
  "packageRules": [
    {
      "matchPackageNames": ["foo/bar"],
      "versionScheme": "regex:^(?<compatibility>.*)-v?(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)?$"
    }
  ]
}
```

Author: I've made the changes you wanted here, and added the `h2` for Using a custom regex versioning scheme as you suggested.
Resolving this conversation. 😄

The "versioning" is different for each package manager, because different package managers use different versioning schemes.
For example, `npm` uses`1.0.0-beta.1` and `pip` uses `1.0.0b1`.

## Why you might need to manually configure versioning

Renovate interprets versions correctly out-of-the-box most of the time.
It's impossible to automatically detect **all** versioning schemes, so sometimes you need to tell the bot what versioning scheme it should use.

You can manually configure/override the `versioning` value for a particular dependency.
You generally won't need to override the defaults for ecosystems which enforce a strict version scheme like `npm`.

> Configuring or overriding the default `versioning` can be particularly helpful for ecosystems like Docker/Kubernetes/Helm, where versioning is barely a "convention".

Reviewer: ```suggestion
Configuring or overriding the default `versionScheme` can be particularly helpful for ecosystems like Docker/Kubernetes/Helm, where versioning is barely a "convention".
```

## General concepts behind overriding versioning

> - Although you can reconfigure versioning per-manager or per-datasource, you are unlikely to ever to need to do this

Reviewer: ```suggestion
- Although you can reconfigure versioning per-manager or per-datasource, it's unlikely that such a broad change would ever be needed
```

> - More commonly you would need to configure `versioning` for individual packages or potentially package patterns

Reviewer: ```suggestion
- More commonly you would need to configure `versionScheme` for individual packages or potentially package patterns
```

- The best way to do this is with `packageRules`, with a combination of `matchManagers`, `matchDatasources`, `matchPackageNames` and `matchPackagePatterns`

## Examples of versioning overrides

### Overriding Docker versioning to use a versioning specific for a package

> 

Reviewer: ```suggestion

The below package rule overrides the default `docker` versioning to instead use `pep440` when the Docker image is `python`:

```

}
```

### Linuxserver.io helm-values versioning

If you're using [linuxserver.io](https://www.linuxserver.io/) and want to use their versioning, do the following:

```json
{
  "packagePatterns": ["(^|/)linuxserver\\/ddclient$"],
  "packageNames": ["ddclient"],
  "versioning": "loose",
  "allowedVersions": "/^(?<version>v\\d+\\.\\d+\\.\\d+-ls\\d+)$/"
}
```

ALTERNATIVE:

> If you're using [linuxserver.io](https://www.linuxserver.io/) and want to use their versioning, you can use the config below.

Reviewer: We don't want to include such specific information here at the top of the versioning modules page, otherwise the expectation will be "why didn't you include my example too?" and we'd end up with pages of example JSON. It's enough that people (a) are alerted to the fact they may need  to explicitly configure versioning, (b) get some brief examples of doing it, (c) create a Discussion if they want further help

Author: Oh okay, I was thinking maybe you did want to cover this specific use-case.
It's gone in the next commit. 😄

}

> ```

Reviewer: Sorry I think this is now a duplicate of the one I added earlier

Author: Yeah we have a duplicate now. 😄

Author: What heading do you like best:

- `Docker versioning`
- `Overriding Docker versioning to use a versioning specific for a package`

Author: I've picked the more descriptive title to keep.
